android/tangram/src/com/mapzen/tangram/MapData.java

```diff
@@ -1,161 +1,154 @@
 
-    protected MapData(long cPtr, boolean cMemoryOwn) {
-        super(TangramJNI.MapData_SWIGSmartPtrUpcast(cPtr), true);
-        swigCMemOwnDerived = cMemoryOwn;
-        swigCPtr = cPtr;
-    }
+    String name;
+    long pointer = 0;
+    MapController map;
+    List<String> geojson = new ArrayList<>();
+    List<Geometry> features = new ArrayList<>();
 
-    protected static long getCPtr(MapData obj) {
-        return (obj == null) ? 0 : obj.swigCPtr;
+    /**
+     * Construct a collection of drawable map features.
+     * @param name The name of the data collection. Once added to a map, features from this
+     * {@code MapData} will be available from a data source with this name, just like a data source
+     * specified in a scene file.
+     */
+    public MapData(String name) {
+        this.name = name;
     }
 
-    protected void finalize() {
-        delete();
+    /**
+     * Get the name of this {@code MapData}.
+     * @return The name.
+     */
+    public String name() {
+        return name;
     }
 
-    public synchronized void delete() {
-        if (swigCPtr != 0) {
-            if (swigCMemOwnDerived) {
-                swigCMemOwnDerived = false;
-                TangramJNI.delete_MapData(swigCPtr);
-            }
-            swigCPtr = 0;
+    /**
+     * Add the features from this {@code MapData} to a map.
+     * <p>
+     * This {@code MapData} will be associated with the given map until {@link #removeFromMap()}
+     * is called or until this method is called again.
+     * @param map The {@code MapController} managing the destination map.
+     */
+    public void addToMap(MapController map) {
+        if (map == null) {
+            throw new RuntimeException("Cannot assign a null map, use removeFromMap() instead");
+        }
+        if (this.map != null) {
+            removeFromMap();
         }
-        super.delete();
+        this.map = map;
+        this.pointer = map.nativeAddDataSource(name);
+        syncWithMap();
     }
 
-
     /**
-     * Construct a new MapData object for adding drawable data to the map
-     * @param name Name of the data source in the scene file for styling this
-     *             object's data
+     * Remove this {@code MapData} from the map it is currently associated with. This must not be
+     * called when this {@code MapData} is not associated with a map.
      */
-    public MapData(String name) {
-        this(TangramJNI.new_MapData(name, ""), true);
+    public void removeFromMap() {
+        if (map == null) {
+            throw new RuntimeException("There is no associated map for this MapData");
+        }
+        map.nativeRemoveDataSource(pointer);
+        pointer = 0;
+        map = null;
     }
 
     /**
-     * Remove all data from this source
-     * @return This object, for chaining
+     * Synchronize the features in this {@code MapData} with the map.
+     * <p>
+     * This method must be called after features are added or removed to update their
```

Comment:
Yeah, I've also thought about just doing the synchronization automatically whenever a feature is added. The benefit of making this an explicit call is that we can then make the feature data processing much more efficient without changing the interface. For example, if we synchronize 500 features all at once instead of one at a time, then we could process the features with geojsonvt *once* instead of *500 times*.